### PR TITLE
VST: LockBit (longform)

### PR DESCRIPTION
**Pull request by gerbsec**

### CODE REVIEW

**Comment by gerbsec** [comment]
No ATT&CK module here.

**Comment by bfuzzy1** [comment]
There should be some sort of relevancy check as part of this test. Is the endpoint Windows? If not, exit with test passed.

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	Endpoint "github.com/preludeorg/test/endpoint"
+)
+
+var killchain = map[string]string{
+	"T1082": "systeminfo",
+	"T1489": "Stop-Process %s",
+	"T1485": "Remove-Item .\\one.txt",
+}
+
+func test() {
```
Comment (bfuzzy1) [inline tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:21]: Some additions to consider from the CISA report:


Network Service Discovery | T1046 | LockBit 3.0 uses SoftPerfect Network Scanner to scan target networks.

System Information Discovery | T1082 | LockBit 3.0 will enumerate system information to include hostname, host configuration, domain information, local drive configuration, remote shares, and mounted external storage devices.

System Location Discovery: System Language Discovery | T1614.001 | LockBit 3.0 will not infect machines with language settings that match a defined exclusion list.

Data Encrypted for Impact | T1486 | LockBit 3.0 encrypts data on target systems to interrupt availability to system and network resources.

Exfiltration Over Web Service: Exfiltration to Cloud Storage | T1567.002 | LockBit 3.0 actors use (1) rclone, an open source command line cloud storage manager to exfiltrate and (2) MEGA, a publicly available file sharing service for data exfiltration. **(This I think could be as simple as sending a hello packet to MEGA)**

**Review by bfuzzy1 — CHANGES_REQUESTED** [review]

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	"fmt"
+	"net"
+	"os/exec"
+	"strings"
+	"sync"
+
+	Endpoint "github.com/preludeorg/test/endpoint"
+)
+
+var killchain = map[string]string{
```
Comment (bfuzzy1) [inline tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:19]: Just need to add encryption and then use T1485 to remove the files that are created for the encryption process. Make sure to look into how exactly Lockbit does their encryption. Are they encrypting the whole file? The beginning  or a specific amount of bytes, etc.

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+				file.Close()
+				os.Rename("important.txt", "important.txt.lockbit")
+				println("[+] sysinfo")
+				command := append(powershell, value)
+				Endpoint.Shell(command)
+			}
+		}
+		println("[+] Deletion")
+		Endpoint.Remove("important.txt.lockbit")
+		Endpoint.Stop(101)
```
Comment (bfuzzy1) [inline tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:106]: @gerbsec this needs a test passed if any of the tasks are prevented.

Comment (bfuzzy1) [inline tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:106]: @gerbsec this needs a test passed if any of the tasks are prevented.

**Review by bfuzzy1 — CHANGES_REQUESTED** [review]

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
+//go:build windows
+// +build windows
```
Comment (privateducky) [inline tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:2]: can you verify this works/doesn't cause any error logs on the backend? 
@packplusplus or @makk94 can help

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+/*
+ID: 194c2c83-4623-40a2-91c0-9c82745d4235
+NAME: LockBit
+CREATED: 2023-05-01
```
Comment (privateducky) [inline tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:7]: let's update these dates to 05/04 (next Monday)

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+	return key, nil
+}
+
+func T1614() bool {
+	println("[+] Language Check")
+	cmd := exec.Command("powershell", "Get-Culture | select -exp Name")
+	output, err := cmd.Output()
+	if err != nil {
+		Endpoint.Stop(100)
```
[the remark this answers is not in the document]
Comment (bfuzzy1) [reply tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:67]: One of 2 relevancy checks. The first is the Windows build flag then LockBit will check system languages and if they are Russian they will not do anything on them.

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/README.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+## How
+
+> Safety: the test only performs non-destructive tasks and does not attempt to modify the system in any harmful way.
+
+This test involves several stages of scanning and system interaction. First, the test checks the system's language setting and operating system. If the system language is not set to Russian and the operating system is Windows, the test executes a series of PowerShell commands, including:
+* gathering system information 
+* stopping a process 
+* creating and deleting a file 
+* performing file encryption.
+* listing shadow copies
```
[the remark this answers is not in the document]
Comment (bfuzzy1) [reply tests/194c2c83-4623-40a2-91c0-9c82745d4235/README.md:17]: So our thinking is these products don't actually care if the listing of shadow copies actually succeeds or not but rather if a command was executed to list them which we can do and not worry about whether it succeeds or fails in the output.

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go (hunk 5 above)
Comment (bfuzzy1) [reply tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go]: Great question! Need to spend some time thinking about this!

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (privateducky) [reply tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:67]: ack; so you're using PowerShell being available as an indirect proxy for 'if windows'?

the build flag on this test indicates 'windows only', so I don't think we need a separate 'if windows' check... but let's play this out: if this command fails on a windows machine, it means we couldn't get the language setting (for any number of reasons) and we return 100 (PROTECTED). Should that be the case or 104 (not relevant)?

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (privateducky) [reply tests/194c2c83-4623-40a2-91c0-9c82745d4235/README.md:17]: sorry, my question was more: I thought you couldn't list the shadow copies because you were worried it would time out?

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+func T1490() {
+	println("[+] Attempting to read ShadowCopy")
+	command := []string{"powershell.exe", "-c", "Get-WmiObject -Query 'select * from Win32_ShadowCopy'"}
+	Endpoint.Shell(command)
+}
+
+func test() {
+	if T1614() {
```
Comment (privateducky) [inline tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:82]: so if T1614 is false... (language is ru) ... we exit as UNPROTECTED? shouldn't that be PROTECTED?

Comment (privateducky) [reply tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:82]: I'm thinking we need:
```
if T1614() {
...
} else {
   exit(100)
}
exit(101)
```

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+		for key, value := range killchain {
```
Comment (privateducky) [inline tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:83]: this is creative, in the fact that it is aiming to be flexible and create some structure. but in a single test example it's too hard to follow. for example, if there are 3 functions in the 'kill chain' then the else block (i.e., not key == T1489) will get hit twice, which feels odd and unpredictable. 

Instead, just plainly write out the code to what it is doing.

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+			powershell := []string{"powershell.exe", "-c"}
+			if key == "T1489" {
+				println("[+] Killing Process")
+				cmd := exec.Command("powershell.exe", "-c", "sleep", "10")
```
Comment (privateducky) [inline tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:87]: great kill idea; is this 10s though? isn't that too close to the test timeout? why not make it 1s?

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+				err := cmd.Start()
+				if err != nil {
```
Comment (privateducky) [inline tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:89]: I wouldn't have the test exit as 100 anywhere.
this test should either exit as a 104 (not relevant) at an early stage OR the final line of code should be .Stop(101).
any other exit codes should come up because the OS or a defense stopped this from running. we don't want to artificially say the test is protected just because a particular step returned an error.

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (bfuzzy1) [reply tests/194c2c83-4623-40a2-91c0-9c82745d4235/README.md:17]: Oh no the worry was not having admin haha, but that doesn't matter since the command can still be executed.

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go (hunk 9 above)
Comment (privateducky) [reply tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:83]: this one @gerbsec

**Review by bfuzzy1 — APPROVED** [review]

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+func T1490(executor []string) {
+	println("[+] Attempting to read ShadowCopy")
+	command := append(executor, "Get-WmiObject -Query 'select * from Win32_ShadowCopy'")
+	Endpoint.Shell(command)
+}
+
+func T1486() {
+	Endpoint.Write("important.txt", make([]byte, 10000))
```
Comment (privateducky) [inline tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:101]: two things that trigger an EDR on ransomware are 1) frequency of encryption and 2) number of files encrypted.

we should make both of these trivial to adjust, numbers wise. starting off, we should not change the frequency (sleep like 0.1s between each encryption) but we should create/encrypt at least 5 files.

File: tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+	file.WriteAt(encrypted, 0)
+	file.Close()
+	os.Rename("important.txt", "important.txt.lockbit")
+	println("[+] Deleting File")
+	Endpoint.Remove("important.txt.lockbit")
+}
+
+func test() {
+	if T1614() {
+		executor := []string{"powershell.exe", "-c"}
```
Comment (privateducky) [inline tests/194c2c83-4623-40a2-91c0-9c82745d4235/194c2c83-4623-40a2-91c0-9c82745d4235.go:120]: this is interesting because we have 2 approaches we can take on this as well.. 

if we run everything through a single PID (like this) it is easier to detect. if we create different PIDS to do these actions, it becomes harder to detect. I don't think I'd adjust this code, but it is something worth remembering.

**Comment by bfuzzy1** [comment]
Closing per David. This will be under our Security Detect account @gerbsec
